### PR TITLE
DeepSeek V4 Muon supprt

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Add support for Muon optimizer for DeepSeek V4. [This](https://paste.googleplex.com/6071583780241408) was the parameter mapping for parameter -> Muon/AdamW used, following quote from the paper: _"We maintain the AdamW (Loshchilov and Hutter, 2017) optimizer for the embedding module, the prediction head module, the static biases and gating factors of mHC modules, and the weights of all RMSNorm modules. All other modules are updated with Muon."_

# Tests

Can see the the perplexity/lm_loss decreasing in [this run](https://cloudlogging.app.goo.gl/9gkZh1ask98nrUqG6) using [this train command](https://paste.googleplex.com/6612666243219456) where we train with Muon.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/optimizers/optimizers.py [hunk 1]
```diff
@@ -197,16 +197,29 @@ def get_optimizer(config, learning_rate_schedule, model=None):
       muon_weight_dimension_numbers = get_muon_weight_dimension_numbers(model, config)
     else:
       raise ValueError("Please specify model to extract muon dimension number.")
+
+    # pylint: disable=import-outside-toplevel
+    from maxtext.configs.types import DecoderBlockType
```
Comment (shuningjin) [inline src/maxtext/optimizers/optimizers.py]: nit: can we import on top?

File: src/maxtext/utils/muon_utils.py [hunk 2]
```diff
@@ -70,7 +70,11 @@ def transform_logic(path: Tuple[str, ...]) -> Optional[mdn]:
   """
 
   # 1 Exclude parameters not suitable for Muon (scalar, embeddings, unembedding)
-  if _is_path_contain_any(("scale", "bias", "embedding", "logits_dense"), path):
+  if any(
+      any(x in segment for x in ("scale", "embedding", "logits_dense", "beta", "base", "sinks", "tid2eid", "inv_freq"))
+      or segment == "bias"
+      for segment in path
+  ):
```
Comment (shuningjin) [inline src/maxtext/utils/muon_utils.py:98]: https://paste.googleplex.com/6506764068585472

1. I don't see `inv_freq` in model structure: can remove
2. `base` is too generic: can change to `hc_base`
3. `beta` is too generic: maybe use `post_beta`, `pre_beta`, `res_beta`
4. can we combine "bias" into the list?
5. add comments for future note
```
# "embedding": embedding
# "logits_dense": output embedding
# "tid2eid": lookup table in hash routing moe
# "scale": scalar, common module
# "sinks": scalar, attention sink 
# "bias": scalar, common module
# "hc_base": scalar, in mhc head
# "post_beta", "pre_beta", "res_beta": scalar, in mhc
```

File: src/maxtext/utils/muon_utils.py [hunk 3]
```diff
@@ -70,7 +70,11 @@ def transform_logic(path: Tuple[str, ...]) -> Optional[mdn]:
   """
 
   # 1 Exclude parameters not suitable for Muon (scalar, embeddings, unembedding)
```
Comment (shuningjin) [inline src/maxtext/utils/muon_utils.py:72]: I wonder whether we can add a test to ensure scalar (ndim=1) always have mdn None. can be follow-up

File: src/maxtext/utils/muon_utils.py [hunk 4]
```diff
@@ -182,7 +189,16 @@ def get_model_mdn(model_name, scan_layers=True, verbose=False, pure_nnx=False):
       f"scan_layers={scan_layers}",
       "attention=dot_product",
       f"pure_nnx={pure_nnx}",
+      "override_model_config=True",
```
Comment (shuningjin) [inline src/maxtext/utils/muon_utils.py]: do we need this line?

File: tests/unit/optimizers_test.py [hunk 5]
```diff
@@ -621,6 +795,49 @@ def __init__(self, rngs: nnx.Rngs):
     # Check attention out: [0, -2] -> [-1]
     self.assertEqual(result.self_attention.out.kernel, mdn((0, -2), (-1,)))
 
+  def test_muon_ds4_ns_config(self):
```
Comment (shuningjin) [inline tests/unit/optimizers_test.py]: nit: name `test_muon_ds4_ns_config` -> `test_muon_newton_schulz_config`

**Review by shuningjin — COMMENTED** [review]
Greak work! Thanks for incorporating hybrid Newton Schultz iterations and adding parameter transformation for DeepSeek4! 

The [mapping](https://paste.googleplex.com/6360493588611072) looks reasonable. Only minor comments.

**Review by shuningjin — APPROVED** [review]
LGTM. Thank you!

**Review by parambole — APPROVED** [review]
LGTM

**Review by aireenmei — APPROVED** [review]
